quidel_covidtest/tests/test_geo_maps.py

```diff
@@ -1,29 +1,62 @@
+from datetime import datetime
+
 import pandas as pd
 
-from delphi_quidel_covidtest.geo_maps import geo_map
+from delphi_quidel_covidtest.geo_maps import geo_map, add_megacounties
+from delphi_quidel_covidtest.constants import AGE_GROUPS
 
+DATA_COLS = ['totalTest', 'numUniqueDevices', 'positiveTest']
 
 class TestGeoMap:
     def test_county(self):
 
         df = pd.DataFrame(
             {
-                "zip": [1607, 1740, 98661, 76010, 76012, 76016],
-                "timestamp": ["2020-06-15", "2020-06-15", "2020-06-15",
-                              "2020-06-15", "2020-06-15", "2020-06-15"],
-                "totalTest": [100, 50, 200, 200, 250, 500],
-                "positiveTest": [10, 8, 15, 5, 20, 50],
-                "numUniqueDevices": [2, 1, 1, 1, 1, 1]
+                "zip": [1607, 1740, 1001, 1003, 98661, 76010, 76012, 76016],
+                "timestamp": ["2020-06-15", "2020-06-15", "2020-06-15", "2020-06-15",
+                              "2020-06-15", "2020-06-15", "2020-06-15", "2020-06-15"],
+                "totalTest_total": [10, 16, 10, 15, 20, 200, 250, 500],
+                "positiveTest_total": [6, 8, 6, 8, 5, 5, 20, 50],
+                "numUniqueDevices_total": [2, 1, 2, 1, 1, 1, 1, 1],
             }
         )
+        
+        for agegroup in AGE_GROUPS[1:-1]:
+            df[f"totalTest_{agegroup}"] = [2, 3, 2, 2, 4, 40, 20, 25]
+            df[f"positiveTest_{agegroup}"] = [1, 1, 1, 1, 1, 1, 3, 8]
+            df[f"numUniqueDevices_{agegroup}"] = [2, 1, 2, 1, 1, 1, 1, 1]
+        
+        df[f"totalTest_age_0_17"] = [4, 6, 4, 4, 8, 80, 40, 50]
+        df[f"positiveTest_age_0_17"] = [2, 2, 2, 2, 2, 2, 6, 16]
+        df[f"numUniqueDevices_age_0_17"] = [2, 1, 2, 1, 1, 1, 1, 1]
 
         new_df, res_key = geo_map("county", df)
 
         assert res_key == 'fips'
-        assert set(new_df["fips"].values) == set(['25027', '53011', '48439'])
+        assert set(new_df["fips"].values) == set(['25027', '25013', '25015', '53011', '48439'])
         assert set(new_df["timestamp"].values) == set(df["timestamp"].values)
-        assert set(new_df["totalTest"].values)  == set([150, 200, 950])
-        assert set(new_df["positiveTest"].values) == set([18, 15, 75])
+        assert set(new_df["totalTest_total"].values)  == set([26, 10, 15, 20, 950])
+        assert set(new_df["positiveTest_total"].values) == set([14, 6, 8, 5, 75])
+
+        assert set(new_df["totalTest_age_0_4"].values)  == set([5, 2, 2, 85, 4])
+        assert set(new_df["positiveTest_age_0_4"].values) == set([2, 1, 1, 12, 1])
+        
+        assert set(new_df["totalTest_age_0_17"].values)  == set([10, 4, 4, 170, 8])
+        assert set(new_df["positiveTest_age_0_17"].values) == set([4, 2, 2, 24, 2])
+        
+        # Test Megacounties
+        new_df["timestamp"] = [datetime.strptime(x, "%Y-%m-%d") for x in new_df["timestamp"]]
+        mega_df = add_megacounties(new_df, True)
```

Comment:
I noticed that mega_df here looks something like 
```
(Pdb) mega_df[mega_df.columns[:5]]
   timestamp   fips state_id  totalTest_total  positiveTest_total
0 2020-06-15  25013       ma             10.0                 6.0
1 2020-06-15  25015       ma             15.0                 8.0
2 2020-06-15  25027       ma             26.0                14.0
3 2020-06-15  48439       tx            950.0                75.0
4 2020-06-15  53011       wa             20.0                 5.0
0 2020-06-15  25000      NaN             25.0                14.0
1 2020-06-15  53000      NaN             20.0                 5.0
```
Do we want the NAs in state_id there? Probably, since you don't want to double count if you aggregate over state_id later, right? Also, does the value 25.0 for 25000 look right to you? It looks like it just combined the first two FIPS codes together. Just want to make sure, because I'm not entirely sure about the functionality of megacounties here (looks like a form of incomplete aggregation? a bit confused).

Reply:
Thanks Dmitry. I should double check the NAs here. 

As for 25.0, it is correct. We only want to merge the counties with total counts smaller than a certain threshold(25 for smoothed signals, 50 for raw signals) for megacounties

Reply:
The NA problem is fixed. state_id will be used for geo aggregation in `generate_sensor_for_(non)parent_geo`, so it cannot be NAs, thank for noticing that.

Reply:
No problem!

Just to double check: so where do we filter out the counties with a low threshold after we do add_megacounties?

Reply:
In `generate_sensor_for_(non)parent_geo`, we have signal functions from `data_tools` where we have the availability threshold to be 50. All the counties(including the megacounties) do not have enough total counts will be filtered out.

For smoothed signals, we use `_geographical_pooling` where we make the borrow proportion to be 0 if the we borrow too much(more than what we currently have, which means, for smoothed signals, if we have less than 25 counts, we need to borrow more than 25 counts. At that time, `_geographical_pooling` will force the borrow proportion to be 0) fake counts from states. That results in less than 50 counts for all the locations with less than 25 counts. The counties that will be merged into megacounties will definitely be filtered out even aftering the geographical pooling step. That's why I set 25 to be the threshold for smoothed signals when adding megacounties.

Reply:
Actually, do you think you could write a unit test that produces megacounties and:
1) makes sure that below thresholded counties are filtered out and aggregated
2) makes sure that counties above threshold are left alone and not aggregated?

Reply:
For the second one, the current unit test `test_county` in `test_geo_maps` do check whether we have some of counties(counts < 25) merged but the others left (>=25). That's how df is designed.

Reply:
For the first one, we already have unit tests (`test_raw_positive_prop`, `test__geographical_pooling`) for functions in `test_data_tools` to test their functionality including whether the borrow proportion is expected and whether the positive percentage with inadequate total counts is turned into nan.

My logical is to add megacounties and later on all the counties (including megacounties are treated the same way). So the current tests are enough.

Reply:
Let me think about it a little more, I need to understand this a bit better.

Reply:
Alright, after looking at it more I am more inclined to believe that this code does what it says, it just was not easy to follow all the threads of pooling and aggregation that live in many different functions. 

I _would_ still feel more comfortable with a test_run test that at least checks that below-threshold counties aren't output. So I wrote a test like this and ran into unexpected behavior. Could you take a look and see? It's in test_run.py and I left a comment about what I did. Maybe I'm misunderstanding something, but it seems like a county that should